src/python/pants/util/contextutil.py

```diff
@@ -55,16 +55,47 @@ def setenv(key, val):
 
 
 @contextmanager
-def stdio_as(stdout, stderr, stdin=None):
-  """Redirect sys.{stdout, stderr, stdin} to alternate file-like objects."""
-  old_stdout, sys.stdout = sys.stdout, stdout
-  old_stderr, sys.stderr = sys.stderr, stderr
-  if stdin:
-    old_stdin, sys.stdin = sys.stdin, stdin
-  yield
-  sys.stdout, sys.stderr = old_stdout, old_stderr
-  if stdin:
-    sys.stdin = old_stdin
+def _stdio_stream_as(src_fd, dst_fd, dst_sys_attribute, mode):
+  """Replace the given dst_fd and attribute on `sys` with an open handle to the given src_fd."""
+  if src_fd == -1:
+    src = open('/dev/null', mode)
+    src_fd = src.fileno()
+
+  # Capture the python and os level file handles.
+  old_dst = getattr(sys, dst_sys_attribute)
+  old_dst_fd = os.dup(dst_fd)
+  if src_fd != dst_fd:
+    os.dup2(src_fd, dst_fd)
+
+  # Open up a new file handle to temporarily replace the python-level io object, then yield.
+  new_dst = os.fdopen(dst_fd, mode)
+  setattr(sys, dst_sys_attribute, new_dst)
+  try:
+    yield
+  finally:
+    new_dst.close()
+
+    # Restore the python and os level file handles.
+    os.dup2(old_dst_fd, dst_fd)
+    setattr(sys, dst_sys_attribute, old_dst)
+
+
+@contextmanager
+def stdio_as(stdout_fd, stderr_fd, stdin_fd):
+  """Redirect sys.{stdout, stderr, stdin} to alternate file descriptors.
+
+  As a special case, if a given destination fd is `-1`, we will replace it with an open file handle
+  to `/dev/null`.
+
+  NB: If the filehandles for sys.{stdout, stderr, stdin} have previously been closed, it's
+  possible that the OS has repurposed fds `0, 1, 2` to represent other files or sockets. It's
+  impossible for this method to locate all python objects which refer to those fds, so it's up
+  to the caller to guarantee that `0, 1, 2` are safe to replace.
+  """
```

Comment:
may make more sense to do the `try: os.fdopen(fd) except OSError: pass` bit here as an added safeguard over the docstring warning since afaict there should be no side effects?

Reply:
It's not a requirement that the input file descriptors have been closed: as demonstrated in the unit tests, we can safely replace file descriptors and their python objects in any case where the objects aren't going to be used _while_ they are swapped out (even by an "alias").

"Aliasing" here is along the lines of: "I have two python objects wrapping the same os file descriptor, and I try to use one of them while its descriptor is swapped out by `stdio_as`". But as far as I know, there is no way to detect this type of aliasing (...except in rust =P).

Reply:
rgr